Bottom: 11b07f569ebf19ecf032786c2e2e4c1caf1c19ee
Top:    84d4864d99f15408e57cb36e1a2ca9eabf15e549
Author: Martin Kupec <martin.kupec@kupson.cz>
Date:   2010-12-22 15:54:28 +0100

From 814f972842d2ff071347f5978cca49c356fd223b Mon Sep 17 00:00:00 2001
Subject: [PATCH] Check Net name appropriately


---

diff --git a/src/select.c b/src/select.c
index 94069b0..96d2fb9 100644
--- a/src/select.c
+++ b/src/select.c
@@ -1026,7 +1026,8 @@ SelectObjectByName (int Type, char *Pattern, bool Flag)
         LibraryEntryType *entry;
         ConnectionType conn;
 
-        if (menu->Name && REGEXEC (menu->Name))
+	/* Name[0] and Name[1] are special purpose, not the actual name*/
+        if (menu->Name && REGEXEC (menu->Name + 2))
           {
             for (i = menu->EntryN, entry = menu->Entry; i; i--, entry++)
               if (SeekPad (entry, &conn, false))
